### PR TITLE
query: Added PromQL tests framework supporing multiple stores. First step towars Query pushdown!

### PR DESCRIPTION
**Pull request by bwplotka**
TDD all the way!

Related to https://github.com/thanos-io/thanos/issues/305

## Changes

* Added initial PromQL acceptance tests.
* Improved logging.
* Improved debug matching info.
* Fixed important matching bug.
* Found reset bug, added issue https://github.com/thanos-io/thanos/issues/3644 and commented code.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
Depends on https://github.com/prometheus/prometheus/pull/8300

**Comment by bwplotka** [comment]
Direction is to use promtool unittest, so closing https://github.com/prometheus/prometheus/pull/8300 

Will adjust test sometime, right now probably continuing against my fork first.

**Comment by bwplotka** [comment]
This should be rdy for review 🤗

File: pkg/query/testdata/promql/prometheus/aggregators.test [hunk 1]
```diff
@@ -0,0 +1,499 @@
+load 5m
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/aggregators.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/collision.test [hunk 2]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/collision.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/functions.test [hunk 3]
```diff
@@ -0,0 +1,892 @@
+# Testdata for resets() and changes().
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/functions.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/histograms.test [hunk 4]
```diff
@@ -0,0 +1,193 @@
+# Two histograms with 4 buckets each (x_sum and x_count not included,
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/histograms.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/literals.test [hunk 5]
```diff
@@ -0,0 +1,59 @@
+eval instant at 50m 12.34e6
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/literals.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/operators.test [hunk 6]
```diff
@@ -0,0 +1,469 @@
+load 5m
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/operators.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/selectors.test [hunk 7]
```diff
@@ -0,0 +1,201 @@
+load 10s
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/selectors.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/staleness.test [hunk 8]
```diff
@@ -0,0 +1,51 @@
+load 10s
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/staleness.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/prometheus/subquery.test [hunk 9]
```diff
@@ -0,0 +1,117 @@
+load 10s
```
Comment (bwplotka) [inline pkg/query/testdata/promql/prometheus/subquery.test:1]: No need to review this, it's 1:1 copied from Prometheus.

File: pkg/query/testdata/promql/thanos/aggregators.test [hunk 10]
```diff
@@ -0,0 +1,218 @@
+store {} 0 30m
```
Comment (bwplotka) [inline pkg/query/testdata/promql/thanos/aggregators.test:1]: It's 1:1 copied from Prometheus, except beginning where we load multiple stores.

**Review by bwplotka — COMMENTED** [review]
NOTE: No need to review `*.test` files. Those are copied from Prometheus.

File: pkg/store/prometheus.go [hunk 11]
```diff
@@ -434,25 +429,26 @@ func (p *PrometheusStore) startPromRemoteRead(ctx context.Context, q *prompb.Que
...
+func matchesExternalLabels(ms []storepb.LabelMatcher, externalLabels labels.Labels) (bool, []*labels.Matcher, error) {
+	tms, err := storepb.MatchersToPromMatchers(ms...)
 	if err != nil {
 		return false, nil, err
 	}
 
-	var newMatcher []storepb.LabelMatcher
+	if len(externalLabels) == 0 {
+		return true, tms, nil
+	}
```
Comment (yeya24) [inline pkg/store/prometheus.go:442]: IMO this length check should be before line 435? We don't need to allocate convert matchers in this case.

Comment (bwplotka) [reply pkg/store/prometheus.go:442]: We might need to, as we want to return translated slice.

Comment (yeya24) [reply pkg/store/prometheus.go:442]: You are right. I missed that.

**Review by kakkoyun — APPROVED** [review]
LGTM 🥇 Amazing work.

**Comment by bwplotka** [comment]
Can't believe it was merged so fast! Thanks ❤️

**Pull request merged**